plugin-gestures/src/main/java/com/mapbox/maps/plugin/gestures/GesturesPluginImpl.kt

```diff
@@ -1309,6 +1315,22 @@ class GesturesPluginImpl : GesturesPlugin, GesturesSettingsBase {
     return true
   }
 
+  @VisibleForTesting(otherwise = PRIVATE)
+  internal fun isPointAboveHorizon(
+    pixel: ScreenCoordinate
+  ): Boolean {
+    if (mapProjectionDelegate.getMapProjection() != MapProjection.Mercator) {
+      return false
+    }
+    // Prevent drag start in area around horizon to avoid sharp map movements
+    val topMapMargin = 0.04 * mapTransformDelegate.getSize().height
+    val reprojectErrorMargin = min(10.0, topMapMargin / 2)
+    val point = ScreenCoordinate(pixel.x, pixel.y - topMapMargin)
+    val coordinate = mapCameraManagerDelegate.coordinateForPixel(point)
+    val roundtripPoint = mapCameraManagerDelegate.pixelForCoordinate(coordinate)
```

Comment:
The essential part of the approach is drag implementation that's using handleMove to start the drag - so, even when the drag start origin is above e.g. mountain peak, once it gets inside ground (terrain or horizon area), eventually it get's picked up and `dragInProgress` gets set.

E.g. second 16 of the attached video, drag starts when entering mountain peak area that is above horizon.

https://user-images.githubusercontent.com/549216/144406878-b21e3f49-c248-4b96-a6ed-5cb943a269d1.mp4

Roundtrip approach here uses arbitrary reprojectErrorMargin = 10 of 10 pixels. It is approach similar to CPU side occlusion approach for gl-js HTML web markers or Mobile SDK annotations, where original location (A) is projected to screen and then de-projected (including raycast, in case of terrain) back to map (location B). If two locations A and B are further than some threshold, then the point is occluded or above horizon.

![Aug-14-2020 16-06-32](https://user-images.githubusercontent.com/549216/90252440-2cde5000-de48-11ea-8f3f-6bffa52e07e1.gif)

Something additional on handleMove and arbitrary reproject error margin. The value of 10 pixels is experimental, large enough to work in all platforms tested. Even if there is a theoretical possibility that raycast could hit missing elevation tile or border between DEM tiles border, leading to potential offset higher than 10 px on e.g. 4K screen (couldn't reproduce it though), handleMove will eventually get dragInProgress triggered.

Finally, to answer the question, this is the performant and terrain independent way (works for terrain and 2D) to detect if a point is above horizon or above mountain peak that is above horizon line. This is because the approach in both cases clamps to topmost horizon or terrain point.

```
    val coordinate = mapCameraManagerDelegate.coordinateForPixel(point)
    val roundtripPoint = mapCameraManagerDelegate.pixelForCoordinate(coordinate)
```

https://github.com/mapbox/mapbox-gl-js/blob/941ba0fdb8d1b945dcd7040c129bff559ed5a5e5/src/geo/transform.js#L1205-L1212

Reply:
Very interesting. Thanks for the explanation.